RoaringBitmap/src/test/java/org/roaringbitmap/art/Node16Test.java

```diff
@@ -0,0 +1,306 @@
+
+    // check that searching for each key, is FOUND
+    for (int i = 0; i < insertCount; i++) {
+      byte key = (byte) (i + keyOffset);
+      SearchResult sr = nodes.getNearestChildPos(key);
+
+      Assertions.assertEquals(SearchResult.Outcome.FOUND, sr.outcome);
+      Assertions.assertTrue(sr.hasKeyPos());
+      // the positions are zero based, even though the keys values are offset
+      Assertions.assertEquals(i, sr.getKeyPos());
+      Assertions.assertEquals(key, nodes.getChildKey(sr.getKeyPos()));
+    }
+
+    // search before the first value "keyOffset", and surprise, nothing will be found
+    SearchResult sr = nodes.getNearestChildPos((byte) (keyOffset - 1));
+    Assertions.assertEquals(SearchResult.Outcome.NOT_FOUND, sr.outcome);
+    Assertions.assertFalse(sr.hasKeyPos());
+    Assertions.assertEquals(Node.ILLEGAL_IDX, sr.getNextSmallerPos());
+
+    // search after the last value aka "insertCount", and surprise, nothing will be found
+    sr = nodes.getNearestChildPos((byte) (keyOffset + insertCount));
+    Assertions.assertEquals(SearchResult.Outcome.NOT_FOUND, sr.outcome);
+    Assertions.assertFalse(sr.hasKeyPos());
+    Assertions.assertEquals(Node.ILLEGAL_IDX, sr.getNextLargerPos());
+  }
+
+  @Test
+  public void testSparseNonZeroBasedKeysSearch() {
+    Node nodes = new Node16(0);
+    final int insertCount = 15;
+    final int lastValue = insertCount - 1;
+
+    final int step = 3;
+    final int keyOffset = 0x20;
+
+    // create the data
+    for (int i = 0; i < insertCount; i++) {
+      LeafNode leafNode = new LeafNode(i, i);
+      byte key = (byte) ((i * step) + keyOffset);
+      nodes = Node16.insert(nodes, leafNode, key);
+    }
+    // check we are testing the correct thing
+    Assertions.assertTrue(nodes instanceof Node16);
+
+    // check that searching for each key, is FOUND
+    for (int i = 0; i < insertCount; i++) {
+      byte key = (byte) ((i * step) + keyOffset);
+      SearchResult sr = nodes.getNearestChildPos(key);
+
+      Assertions.assertEquals(SearchResult.Outcome.FOUND, sr.outcome);
+      Assertions.assertTrue(sr.hasKeyPos());
+      // the positions are zero based, even though the keys values are offset
+      int keyPos = sr.getKeyPos();
+      Assertions.assertEquals(i, keyPos);
+      Assertions.assertEquals(key, nodes.getChildKey(sr.getKeyPos()));
+
+      // search in the "gaps" before the key
+      {
+        byte bKey = (byte)(key - 1);
+        sr = nodes.getNearestChildPos(bKey);
+        Assertions.assertEquals(SearchResult.Outcome.NOT_FOUND, sr.outcome);
+        Assertions.assertFalse(sr.hasKeyPos());
+
+        // the value smaller than the first should be INVALID, and the rest should be the prior key
+        if (i == 0) {
+          Assertions.assertEquals(Node.ILLEGAL_IDX, sr.getNextSmallerPos());
+        } else {
+          Assertions.assertEquals((((i - 1) * step) + keyOffset), Byte.toUnsignedInt(nodes.getChildKey(sr.getNextSmallerPos())));
```

Comment:
Maybe assign the two side to variables first. Long lines and complicated expressions are hard to read and to verify that it's asserting the right thing.